istioctl/pkg/writer/pilot/status.go

```diff
@@ -29,110 +29,222 @@ import (
+	return types
+}
+
+// setupStatusPrint processes the discovery responses and prepares them for printing
+func (s *XdsStatusWriter) setupStatusPrint(drs map[string]*discovery.DiscoveryResponse) (*tabwriter.Writer, []*xdsWriterStatus, []string, error) {
 	var fullStatus []*xdsWriterStatus
-	mappedResp := map[string]string{}
+	allTypesSet := map[string]struct{}{}
 	w := new(tabwriter.Writer).Init(s.Writer, 0, 8, 5, ' ', 0)
-	_, _ = fmt.Fprintln(w, "NAME\tCLUSTER\tCDS\tLDS\tEDS\tRDS\tECDS\tISTIOD\tVERSION")
+
+	// Process each discovery response
 	for _, dr := range drs {
 		for _, resource := range dr.Resources {
 			clientConfig := xdsstatus.ClientConfig{}
-			err := resource.UnmarshalTo(&clientConfig)
-			if err != nil {
-				return nil, nil, fmt.Errorf("could not unmarshal ClientConfig: %w", err)
+			if err := resource.UnmarshalTo(&clientConfig); err != nil {
+				return nil, nil, nil, fmt.Errorf("could not unmarshal ClientConfig: %w", err)
 			}
+
 			meta, err := model.ParseMetadata(clientConfig.GetNode().GetMetadata())
 			if err != nil {
-				return nil, nil, fmt.Errorf("could not parse node metadata: %w", err)
+				return nil, nil, nil, fmt.Errorf("could not parse node metadata: %w", err)
 			}
+
+			// Skip if namespace filter is set and doesn't match
 			if s.Namespace != "" && meta.Namespace != s.Namespace {
 				continue
 			}
-			cds, lds, eds, rds, ecds := getSyncStatus(&clientConfig)
-			cp := multixds.CpInfo(dr)
-			fullStatus = append(fullStatus, &xdsWriterStatus{
-				proxyID:               clientConfig.GetNode().GetId(),
-				clusterID:             meta.ClusterID.String(),
-				istiodID:              cp.ID,
-				istiodVersion:         meta.IstioVersion,
-				clusterStatus:         cds,
-				listenerStatus:        lds,
-				routeStatus:           rds,
-				endpointStatus:        eds,
-				extensionconfigStatus: ecds,
-			})
-			if len(fullStatus) == 0 {
-				return nil, nil, fmt.Errorf("no proxies found (checked %d istiods)", len(drs))
+
+			// Process XDS configs
+			configs := handleAndGetXdsConfigs(&clientConfig)
+			typeStatus := make(map[string]string, len(configs))
+			for _, config := range configs {
+				typeURL := config.GetTypeUrl()
+				allTypesSet[typeURL] = struct{}{}
+				typeStatus[typeURL] = formatStatus(config)
 			}
 
-			sort.Slice(fullStatus, func(i, j int) bool {
-				return fullStatus[i].proxyID < fullStatus[j].proxyID
+			fullStatus = append(fullStatus, &xdsWriterStatus{
+				proxyID:       clientConfig.GetNode().GetId(),
+				clusterID:     meta.ClusterID.String(),
+				istiodID:      multixds.CpInfo(dr).ID,
+				istiodVersion: meta.IstioVersion,
+				typeStatus:    typeStatus,
 			})
 		}
 	}
-	if len(mappedResp) > 0 {
-		mresp, err := json.MarshalIndent(mappedResp, "", "  ")
-		if err != nil {
-			return nil, nil, err
+
+	if len(fullStatus) == 0 {
+		return nil, nil, nil, fmt.Errorf("no proxies found (checked %d istiods)", len(drs))
```

Comment:
Wanted to call out that there's a behavior change here which breaks foreign tooling calling `istioctl`, not sure if it was intentional or not. Before this patch:
```
istioctl version && istioctl proxy-status && echo $?
client version: 1.26.2
control plane version: 1.26.2
data plane version: none
NAME     CLUSTER     CDS     LDS     EDS     RDS     ECDS     ISTIOD     VERSION
0
```

With this patch:
```
./istioctl version && ./istioctl proxy-status || echo $?
client version: 1.27.0
control plane version: 1.26.2
data plane version: none
Error: failed to setup status print: no proxies found (checked 1 istiods)
1
```

Reply:
Thanks for calling this out - @AritraDey-Dev can you patch?